src/kernels/MIOpenDropoutHIP.cpp

```diff
@@ -0,0 +1,374 @@
+                                     int in_str1,
+                                     int in_str2,
+                                     int in_str3,
+                                     uchar* reserveSpace,
+                                     size_t total_work,
+                                     size_t in_offset,
+                                     size_t out_offset,
+                                     size_t rsvsp_offset)
+{
+
+    dropoutfw<_FLOAT, READ_DAT_TYPE, READ_BOOL_TYPE, USE_MASK, USE_RSVSP>(state,
+                                                                          dropout,
+                                                                          scale,
+                                                                          dim1,
+                                                                          dim2,
+                                                                          dim3,
+                                                                          dim4,
+                                                                          y,
+                                                                          out_str0,
+                                                                          out_str1,
+                                                                          out_str2,
+                                                                          out_str3,
+                                                                          x,
+                                                                          in_str0,
+                                                                          in_str1,
+                                                                          in_str2,
+                                                                          in_str3,
+                                                                          reserveSpace,
+                                                                          total_work,
+                                                                          in_offset,
+                                                                          out_offset,
+                                                                          rsvsp_offset);
+}
+#endif
+
+// Dropout Backward
+#if !RUN_FORWARD
+
+#ifndef USE_PRNG
+#define USE_PRNG 0
+#endif
+
+/**
+ * @brief Applies dropout backward operation to the input data.
+ * The mask is either generated or an existing mask is loaded into
+ * registers from the reserveSpace. The loaded mask is applied to the data in the registers. The
+ * processed data is then written to the output. Reading and writing of input data can be performed
+ * as vectorized operations, with the extent of vectorization determined by #RD_BLOCK.
+ */
+template <typename F, typename T, typename B, bool PRNG = false>
+__forceinline__ __device__ void dropoutbw(const rocrand_state_xorwow* state,
```

Comment:
seems like `dropoutfw` and `dropoutbw` are virtually the same function and all that matters is swapping pointers and strides?

Reply:
Yes, they are. Well, I personally think having more #defines or if constexprs within the kernel ruins the readability. Does this cause any other issues other than MIOpen needing extra space to inline the kernel src?

Reply:
It depends. Probably you don't even need ifdefs and just swap strides and tensors on a host side.

Reply:
@sgundabo : when 90% of the code is the same, we should look to remove duplication. Sure, slight more effort to refactor and read, but much better in the long term as you are not maintaining two copies of the code.

Reply:
I think we can do it later.